src/utils/utils.jl

```diff
@@ -115,57 +116,77 @@ function validate_exported_names(mod::Module)
     return is_valid
 end
 
+# Make match_fn optional
+compare_values(x, y; kwargs...) = compare_values(nothing, x, y; kwargs...)
+
+# Get the default match_fn if necessary. Only call when you know you're done recursing
+_fetch_match_fn(match_fn::Function) = match_fn
+_fetch_match_fn(::Nothing) = isequivalent
+
+# Whether to stop recursing and apply the match_fn
+_is_compare_directly(::DataType, ::DataType) = true
+_is_compare_directly(::T, ::U) where {T, U} = true
+_is_compare_directly(::T, ::T) where {T} = isempty(fieldnames(T))
+
 """
 Recursively compares struct values. Prints all mismatched values to stdout.
 
 # Arguments
-
+  - `match_fn`: optional, a function used to determine whether two values match in the base
+    case of the recursion. If `nothing` or not specified, the default implementation uses
+    `IS.isequivalent`.
   - `x::T`: First value
-  - `y::T`: Second value
+  - `y::U`: Second value
   - `compare_uuids::Bool = false`: Compare any UUID in the object or composed objects.
   - `exclude::Set{Symbol} = Set{Symbol}(): Fields to exclude from comparison. Passed on
      recursively and so applied per type.
 """
-function compare_values(
-    x::T,
-    y::T;
-    compare_uuids = false,
-    exclude = Set{Symbol}(),
-) where {T}
+function compare_values(match_fn::Union{Function, Nothing}, x::T, y::U;
+    compare_uuids = false, exclude = Set{Symbol}()) where {T, U}
+    # Special case: if match_fn is nothing, try calling the two-argument version to maintain
+    # backwards compatibility with packages that only implement that. Keep track of this to
+    # avoid infinite recursion. TODO remove in next major version
+    if isnothing(match_fn) && !(COMPARE_VALUES_SENTINEL in exclude)
+        return compare_values(x, y; compare_uuids = compare_uuids,
+            exclude = union(exclude, [COMPARE_VALUES_SENTINEL]))
+    end
+    exclude = setdiff(exclude, [COMPARE_VALUES_SENTINEL])
```

Comment:
This is a little hacky but I am convinced it is the least bad way to accomplish the task without changing the interface at all (and it's temporary)

Reply:
(ask me what I learned about `Core.kwcall` today)

Reply:
lol ...